tests/common.go

```diff
@@ -552,6 +552,71 @@ func ValidateContext(ctx *scheduler.Context, errChan ...*chan error) {
 	})
 }
 
+func ValidatePureCloudDriveTopologies() error {
+	nodes, err := Inst().V.GetDriverNodes()
+	if err != nil { return err }
+	nodesMap := node.GetNodesByName()
+
+	driverNamespace, err := Inst().V.GetVolumeDriverNamespace()
+	if err != nil { return err }
+
+	pxPureSecret, err := pureutils.GetPXPureSecret(driverNamespace)
+	if err != nil { return err }
+
+	endpointToZoneMap := pxPureSecret.GetArrayToZoneMap()
+	if len(endpointToZoneMap) == 0 {
+		return fmt.Errorf("parsed px-pure-secret endpoint to zone map, but no arrays in map (len==0)")
+	}
+
+	log.Infof("Endpoint to zone map: %v", endpointToZoneMap)
+
+	for _, node := range nodes {
+		log.Infof("Inspecting drive sets on node %v", node.SchedulerNodeName)
+		nodeFromMap, ok := nodesMap[node.SchedulerNodeName]
+		if !ok {
+			return fmt.Errorf("Failed to find node %s in node map", node.SchedulerNodeName)
+		}
+
+		var nodeZone string
+		if nodeFromMap.SchedulerTopology != nil && nodeFromMap.SchedulerTopology.Labels != nil {
+			if z, ok := nodeFromMap.SchedulerTopology.Labels["topology.portworx.io/zone"]; ok {
+				nodeZone = z
+			}
+		}
+
+		if nodeZone == "" {
+			log.Warnf("Node %s has no zone (missing the topology.portworx.io/zone label), skipping drive set checks for it", node.SchedulerNodeName)
+			continue
+		}
+
+		driveSet, err := Inst().V.GetDriveSet(&nodeFromMap)
+		if err != nil { return err }
+
+		for configID, driveConfig := range driveSet.Configs {
+			err = nil
+			if driveConfig.Labels == nil || len(driveConfig.Labels) == 0 {
+				return fmt.Errorf("drive config %s has no labels: validate that you're running on PX master or 3.0+ and using FlashArray cloud drives with topology enabled", configID)
+			}
+
+			var arrayEndpoint string
+			if arrayEndpoint, ok = driveConfig.Labels[pureutils.CloudDriveFAMgmtLabel]; !ok {
+				return fmt.Errorf("drive config %s is missing the '%s' label: validate that you're running PX master or 3.0+ and using FlashArray cloud drives with topology enabled", configID, pureutils.CloudDriveFAMgmtLabel)
+			}
+
+			var driveZone string
+			if driveZone, ok = endpointToZoneMap[arrayEndpoint]; !ok {
+				return fmt.Errorf("drive config %s is on array with endpoint '%s', which is not listed in px-pure-secret", configID, arrayEndpoint)
+			}
+
+			if driveZone != nodeZone {
```

Comment:
It looks to me we can only run positive tests today and we can only use Jenkins to test negative tests. In our test plan confluence wiki, please add one section to describe how we do the negative tests, so we know how to add/modify negative tests when needed.

Reply:
Added an "additional testing notes" section here: https://portworx.atlassian.net/wiki/spaces/PE/pages/2596339877/FACD+CSI+Topology+Test+Plan